internal/provider/resource_apply_module.go

```diff
@@ -0,0 +1,573 @@
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// Configure lets the provider implement the ResourceWithConfigure interface.
+func (t *applyModuleResource) Configure(_ context.Context,
+	req resource.ConfigureRequest, _ *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+	t.client = req.ProviderData.(*tharsis.Client)
+}
+
+func (t *applyModuleResource) Create(ctx context.Context,
+	req resource.CreateRequest, resp *resource.CreateResponse) {
+
+	// Retrieve values from apply module.
+	var applyModule ApplyModuleModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &applyModule)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Do plan and apply, no destroy.
+	var didRun doRunOutput
+	resp.Diagnostics.Append(t.doRun(ctx, &doRunInput{
+		model: &applyModule,
+	}, &didRun)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Update the plan with the computed ID.
+	applyModule.ID = types.StringValue(uuid.New().String())
+	applyModule.ModuleVersion = types.StringValue(didRun.moduleVersion)
+
+	// Set the response state to the fully-populated plan, whether or not there is an error.
+	resp.Diagnostics.Append(resp.State.Set(ctx, applyModule)...)
+}
+
+func (t *applyModuleResource) Read(ctx context.Context,
+	req resource.ReadRequest, resp *resource.ReadResponse) {
+
+	// Get the current state.
+	var state ApplyModuleModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var appliedModuleInfo appliedModuleInfo
+	resp.Diagnostics.Append(t.getCurrentApplied(ctx, state, &appliedModuleInfo)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Update the state with the computed module source and version.
+	if appliedModuleInfo.moduleSource != nil {
+		state.ModuleSource = types.StringValue(*appliedModuleInfo.moduleSource)
+	} else {
+		state.ModuleSource = types.StringNull()
+	}
+	if appliedModuleInfo.moduleVersion != nil {
+		state.ModuleVersion = types.StringValue(*appliedModuleInfo.moduleVersion)
+	} else {
+		state.ModuleVersion = types.StringNull()
+	}
+
+	// TODO: Eventually, when the API and SDK support speculative plan runs with a module source,
+	// this should do a speculative plan run here to determine whether changes are needed.
+
+	// Set the refreshed state, whether or not there is an error.
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+func (t *applyModuleResource) Update(ctx context.Context,
+	req resource.UpdateRequest, resp *resource.UpdateResponse) {
+
+	// Retrieve values from plan.
+	var plan ApplyModuleModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// TODO: Please note that when the API and SDK support speculative runs with a module source,
+	// this will need to look at the results from the Read method's speculative run to determine
+	// whether to do an update.  A way will have to be found to force Terraform to allow the update.
```

Comment:
If we aren't thinking to do the plan in the plan/read stage and finishing the apply in the create/update stage, then I think we just need to add module support for speculative runs.

Reply:
I agree that we should support speculative plans for module sources. We'll need to add a story to the backlog for it.

Reply:
I've created a ticket for this.

Reply:
I had been told that speculative runs were out-of-scope for the story behind this MR.  Has that changed?